lib/matplotlib/_constrained_layout.py

```diff
@@ -83,181 +74,350 @@ def do_constrained_layout(fig, renderer, h_pad, w_pad,
+
+        _make_margin_suptitles(fig, renderer, h_pad=h_pad, w_pad=w_pad)
+
+        # if a layout is such that a columns (or rows) margin has no
+        # constraints, we need to make all such instances in the grid
+        # match in margin size.
+        _match_submerged_margins(fig)
+
+        # update all the variables in the layout.
+        fig._layoutgrid.update_variables()
+
+        if _check_no_collapsed_axes(fig):
+            _reposition_axes(fig, renderer, h_pad=h_pad, w_pad=w_pad,
+                             hspace=hspace, wspace=wspace)
         else:
-            cbook._warn_external('constrained_layout not applied.  At least '
-                                 'one axes collapsed to zero width or height.')
+            cbook._warn_external('constrained_layout not applied because '
+                                 'axes sizes collapsed to zero.  Try making '
+                                 'figure larger or axes decorations smaller.')
+        _reset_margins(fig)
 
 
-def _make_ghost_gridspec_slots(fig, gs):
+def _check_no_collapsed_axes(fig):
     """
-    Check for unoccupied gridspec slots and make ghost axes for these
-    slots...  Do for each gs separately.  This is a pretty big kludge
-    but shouldn't have too much ill effect.  The worst is that
-    someone querying the figure will wonder why there are more
-    axes than they thought.
+    Check that no axes have collapsed to zero size.
     """
-    nrows, ncols = gs.get_geometry()
-    hassubplotspec = np.zeros(nrows * ncols, dtype=bool)
-    axs = []
+    for panel in fig.panels:
+        ok = _check_no_collapsed_axes(panel)
+        if not ok:
+            return False
+
     for ax in fig.axes:
-        if (hasattr(ax, 'get_subplotspec')
-                and ax._layoutbox is not None
-                and ax.get_subplotspec().get_gridspec() == gs):
-            axs += [ax]
-    for ax in axs:
-        ss0 = ax.get_subplotspec()
-        hassubplotspec[ss0.num1:(ss0.num2 + 1)] = True
-    for nn, hss in enumerate(hassubplotspec):
-        if not hss:
-            # this gridspec slot doesn't have an axis so we
-            # make a "ghost".
-            ax = fig.add_subplot(gs[nn])
-            ax.set_visible(False)
-
-
-def _make_layout_margins(ax, renderer, h_pad, w_pad):
+        if hasattr(ax, 'get_subplotspec'):
+            gs = ax.get_subplotspec().get_gridspec()
+            lg = gs._layoutgrid
+            if lg is not None:
+                for i in range(gs.nrows):
+                    for j in range(gs.ncols):
+                        bb = lg.get_inner_bbox(i, j)
+                        if bb.width <= 0 or bb.height <= 0:
+                            return False
+    return True
+
+
+def _get_margin_from_padding(object, *, w_pad=0, h_pad=0,
+                             hspace=0, wspace=0):
+
+    ss = object._subplotspec
+    gs = ss.get_gridspec()
+    lg = gs._layoutgrid
+
+    if hasattr(gs, 'hspace'):
```

Comment:
I'm not sure why hspace is public on GridSpec but private on GridSpecFromSubplotSpec, but I'd rather just make it public in both (possibly a property on the latter if we want to keep it readonly) rather than having to do this kinds of checks here.

Reply:
Thats fine with me.  I don't understand it either.  Is there an action item for me here?